### PR TITLE
doc: Add `near-network` reference level documentation `Overview`

### PR DESCRIPTION
**Pull request by pmnoxx**
This is part of `NetworkDocumentation.md` it covers:
- Overview
- Code structure
- Code flow - initiation

See https://github.com/near/NEPs/pull/274 for full PR

### CODE REVIEW

**Comment by pmnoxx** [comment]
@mm-near @bowenwang1996 I'll add more PRs for different sections of the documentation.
If you prefer I can put all of them in one commit, just let me know what works better for you.

**Comment by pmnoxx** [comment]
Other parts of PR are in 

- [x] https://github.com/near/NEPs/pull/281
- [x] https://github.com/near/NEPs/pull/282
- [x] https://github.com/near/NEPs/pull/283
- [x] https://github.com/near/NEPs/pull/285
- [x] https://github.com/near/NEPs/pull/287
- [x] https://github.com/near/NEPs/pull/288

**Review by bowenwang1996 — COMMENTED** [review]
I think `network.md` would be a better name for the file

File: docs/near_network.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# 1. Overview
+Near Protocol uses its own implementation of a custom peer-to-peer network
+Peers who join network are represented by nodes and connected between them by edges.
+
+The purpose of this document is to describe inner workings of `near-network` package;
+and to be used as reference by future engineers to understand network code without any prior knowledge.
```
Comment (matklad) [inline docs/near_network.md:6]: :+1: for stating the purpose&target audience of the doc! It goes a long way towards keeping the thing maintainable over the long run.

**Comment by matklad** [comment]
Let's reflow at 80 columns according to the styleguide: https://github.com/near/nearcore/blob/master/docs/style.md#documentation

File: docs/network.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
+# 1. Overview
+Near Protocol uses its own implementation of a custom peer-to-peer network
+Peers who join network are represented by nodes and connected between them by edges.
```
Comment (mm-near) [inline docs/network.md:3]: s/connected/connections

File: docs/network.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+and therefore was moved to another thread.
+
+Responsibilities:
+- validating edges by checking whenever cryptographic signatures match.
+
+### 2.2 `RoutingTableActor`
+`RoutingTableActor` maintain view of the `P2P network` represented by set of nodes and edges.
+
+In case a message needs to be sent between two nodes, let's call them `A` and `B`.
+If nodes `A` and `B` are connected then the message san be sent directly.
```
Comment (mm-near) [inline docs/network.md:26]: s/san/can

File: docs/network.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+However, if there aren't a node `C` needs to be chosen to forward message from `A` to `B`.
```
Comment (mm-near) [inline docs/network.md:27]: I'd say that: if there is no direct connection, 'RoutingTableActor' is responsible for picking the best path between the two.

File: docs/network.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+- has logic related to exchanging edges between peers
+
+### 2.3 `PeerActor`
+Whenever a new connection gets accepted, an instance of `PeerActor` gets created.
+Each `PeerActor` keeps a physical a `TCP connection` to exactly one peer.
+
+Responsibilities:
+- Maintaining physical connection.
+- Reading messages from peers, decoding them, and then forwarding them to the right place.
+- Encoding messages, writing them to peers on physical layer.
```
Comment (mm-near) [inline docs/network.md:43]: s/writing/sending

File: docs/network.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+- Routing messages between `PeerManagerActor` and other peers/
```
Comment (mm-near) [inline docs/network.md:44]: remove '/'

File: docs/network.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+### 2.4 `PeerManagerActor`
+`PeerManagerActor` is the main actor of `near-network` crate.
+It's acts as a bridge connecting to the world outside, the other peers, and `ClientActor` and `ClientViewActor`, which
+handle processing any operations on the chain.
+`PeerManagerActor` maintains information about p2p network, and indirectly, through `PeerActor`, connections to all other nodes on the network.
```
Comment (mm-near) [inline docs/network.md:50]: maintains information about p2p network (via Routing Table Actor)

Comment (mm-near) [reply docs/network.md:50]: s/to all other nodes/to some nodes

File: docs/network.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+All messages going to other nodes, or coming from other nodes will be routed through this `Actor`.
+`PeerManagerActor` is responsible for accepting incoming connections from the outside world.
```
Comment (mm-near) [inline docs/network.md:52]: (and creating PeerActors to manage them)

File: docs/network.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+- Maintaining list of `PeerActors`, creating, deleting them
+- Routing information about new edges between `PeerActors` and `RoutingTableManager`
+- Routing messages between `ViewClient`, `ViewClientActor` and `PeerActors`, and consequently other peers.
+- Maintains `RouteBack` structure, which has information on how to send replies to messages
+
+# 3. Code flow - initialization
+`PeerManagerActor` actor gets started.
+`PeerManagerActor` open tcp server, which listens to incoming connection.
+It starts `RoutingTableActor`, which then starts `EdgeValidatorActor`.
+When connection incoming connection gets accepted, it starts `PeerActor` on its own thread.
```
Comment (mm-near) [inline docs/network.md:65]: s/it starts/it starts a new

**Review by mm-near — APPROVED** [review]

**Pull request merged**